### PR TITLE
Grab fixes

### PR DESCRIPTION
**Pull request by sethalves**
- improve what happens when someone equips something that already had an action on it
- grab.js now does a grab heart-beat in userData so it doesn't look like an abandoned grab
- things that are non-physical wont receive velocity when they are released
- grab.js calls startDistanceGrab and releaseGrab entity methods and sends grab/release messages to Hifi-Object-Manipulation channel
- action grabs will now send grab messages

### CODE REVIEW

**Comment by sethalves** [comment]
https://app.asana.com/0/26485730907942/95473155131492
https://app.asana.com/0/26485730907942/107567199240584

**Comment by sethalves** [comment]
to test:
- have someone use grab.js to mouse-grab a thing and move it near you
- equip (bumper-grab) the thing and have the other person release their mouse-grab
- unequip the thing and make sure it all looked ok
- also try this with a handControllerGrab.js distance grab

**Comment by hifi-gustavo** [comment]
Build console output for commit: 000d6c8a68d89063648147dc7ce1500208bd71bf available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3186984586b3c40f37c660bf7a6fc0d9ebc3f53b available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8d4545310268ecf9bec4de7a70c9af006544e6d8 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ec554c3c0b5aa828ff8dfd7f7486d6e1264c22e8 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 01c0b4fbc55920afd7ad3f5d05eb6bad916c06cc available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 99ffb6b79d133c2754b340dcbb630a53f22d02eb available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9e53055d439f9336b976182641f1b23d982d9bc5 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9e53055d439f9336b976182641f1b23d982d9bc5 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5a8664052641683e6b905a72660969cdec42d8ff available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 44beca1687edd56c58892963d7ef903b86fffdb2 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b6918f59a693acdcbc40f9bd15beb604756f413f available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: b6918f59a693acdcbc40f9bd15beb604756f413f available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 382a0c48da96c51a1f66e7533f3cf19bf5f6077e available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9db839bc138c3acd01f6ed2b4a4e020772bee8f4 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 91af732516a8613c7f6198dc5ed617bbd534a71b available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d76b8f23358be65174411253308d967d27226f29 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f5be64ed9633d676df6d8ede974179eef165ac31 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0c794cac5d29517037c112df9dc6f0416bf6d464 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3c41df53e457facfa40bdc5249916208dfe1a5e7 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ceda8e9bea199625306e105c5decdb1677ac29a4 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

File: libraries/physics/src/ObjectActionSpring.cpp [hunk 1]
```diff
@@ -37,9 +39,81 @@ ObjectActionSpring::~ObjectActionSpring() {
...
+    int springCount = 0;
+
+    auto ownerEntity = _ownerEntity.lock();
+    if (!ownerEntity) {
+        return false;
+    }
+
+	QList<EntityActionPointer> springDerivedActions;
+    springDerivedActions.append(ownerEntity->getActionsOfType(ACTION_TYPE_SPRING));
+    springDerivedActions.append(ownerEntity->getActionsOfType(ACTION_TYPE_HOLD));
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionSpring.cpp]: Bad indentation at line 65.

File: libraries/physics/src/ObjectActionSpring.cpp [hunk 2]
```diff
@@ -37,9 +39,81 @@ ObjectActionSpring::~ObjectActionSpring() {
...
+    glm::vec3 linearVelocity;
+    glm::vec3 angularVelocity;
+
+    bool valid = false;
+    int springCount = 0;
+
+    auto ownerEntity = _ownerEntity.lock();
+    if (!ownerEntity) {
+        return false;
+    }
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectActionSpring.cpp]: Not a huge deal, but early exits like this should be at the top _before_ approximately 300 bytes of local variables are created.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f80333ee9e010737008896e4dd864a0f301c1cb7 available at: https://gist.github.com/7065e743b3877202a26a62c48d83b53b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7658.dmg

**Comment by AndrewMeadows** [comment]
Ok I think I understand what is going on for some of this PR.  Multiple actions on an object don't play well together (1) so the solution is for each action to consider the settings of all the other similar actions on the same object and compute the average position and then try to achieve that.  This averaging happens for every ObjectActionSpring instance, every Bullet simulation substep.

If this PR fixes grab bugs in the short term I'm willing to mege it, but I think the Action system is getting ripe for an overhaul and this work just makes me more anxious to get that done to prepare for the long term.

In particular: (a) actions currently require locks on the EntityItem inside their updateAction() implementation and I'd like to avoid that pattern, (b) the problem where multiple actions fight over an object's motion is more general than just grab-vs-grab and as we add more fancy actions this problem will just get worse (grab-vs-vehicle, grab-vs-circularOrbit, etc), (c) how we add/remove/delete actions from the physics simulation needs some work so that it is not so fragile and we don't crash on deleted pointers in the future.

(1) Because Bullet will evaluate the actions in the order in which they were created, one at a time.  Later actions may completely override earlier ones.

**Comment by AndrewMeadows** [comment]
This change fixes a bug in the short term so I"m going to merge it.  However, I'll have to eventually fix this in more correctly for the long term.

**Pull request merged**